src/api/v2/handlers.rs

```diff
@@ -26,6 +31,33 @@ pub async fn subscriptions(
 	Ok(SubscriptionId { subscription_id })
 }
 
+pub async fn submit(
+	submitter: Arc<impl transactions::Submit>,
+	transaction: Transaction,
+) -> Result<TransactionHash, Error> {
+	fn base64_decode(data: String) -> Result<Vec<u8>, Error> {
+		general_purpose::STANDARD
+			.decode(data)
+			.map_err(|error| Error::bad_request(format!("Bad Request: {error}")))
+	}
```

Comment:
I think it is better to create a type wrapper for base64 decoding. Like:
```rust
#[derive(Deserialize)]
#[serde(try_from = "String")]
pub struct Base64(pub Vec<u8>);

impl TryFrom<String> for Base64 { ... }
```

Reply:
This is cool, but we want to have custom bad request message in the response if deserialization fails, and instead of handling custom error in rejections it seems there is no other way (but I have to look more on it). That would resolve both having double structure for transaction and make things simpler by avoiding using generics.

Reply:
You can customize Error type for that purpose. I didn't get where are generics here also

Reply:
Its not that easy because warp has default string prefix, which is maybe ok, Ill keep it simple atm and later resolve this issue if needed.

Reply:
`Bad request: {error}` was custom message, now warp is giving `Request body deserialize error: {error}` since more general solution was implemented. I guess both are fine, and I'm ok with setling on this at the moment, but latter we need to think how end users interpret those messages and try to keep consistency. From my perspective, its ok to crash since we don't loose track that something was actually changed. If it bothers, it can be easily be removed.

Reply:
I mean there is status code already, no need to duplicate it, right? So we can rely on defaults.

Reply:
Status code 400 can imply both json parsing and base 64 decoding issue, so message is what make them distinct.

Reply:
Sorry, looked at the code again. LGTM right now! Thanks a lot! :smiley: